labs/Language/IntroTypeHackery.md

```diff
@@ -0,0 +1,257 @@
+>    and _v_ is even, then gcd(_u_, _v_) = gcd(_u_, _v_/2).
+> 4. If _u_ and _v_ are both odd, and _u_ ≥ _v_, then gcd(_u_, _v_) =
+>    gcd((_u_ − _v_)/2, _v_). If both are odd and _u_ < _v_, then
+>    gcd(_u_, _v_) = gcd((_v_ − _u_)/2, _u_). These are combinations
+>    of one step of the simple Euclidean algorithm, which uses
+>    subtraction at each step, and an application of step 3 above. The
+>    division by 2 results in an integer because the difference of two
+>    odd numbers is even.
+> 5. Repeat steps 2–4 until _u_ = _v_, or (one more step) until _u_
+>    = 0. In either case, the GCD is 2<sup>_k_</sup>_v_, where _k_ is
+>    the number of common factors of 2 found in step 2.
+
+[1] Wikipedia contributors. Binary GCD algorithm
+[Internet]. Wikipedia, The Free Encyclopedia; 2020 May 11, 17:27 UTC
+[cited 2020 Jul 10]. Available from:
+[https://en.wikipedia.org/w/index.php?title=Binary_GCD_algorithm&oldid=956130910]().
+
+
+A First Attempt at Binary gcd in Cryptol
+----------------------------------------
+
+```cryptol
+bgcd : {n} (fin n) => [n] -> [n] -> [n]
+bgcd x y = if x == 0 \/ y == 0 then x ^ y
+            | even x /\ even y then bgcd (x >> 1) (y >> 1) << 1
+            | even x           then bgcd (x >> 1) y
+            |           even y then bgcd x (y >> 1)
+            | x >= y           then bgcd ((x - y) >> 1) y
+                               else bgcd x ((y - x) >> 1)
+```
+
+While this computes correct values, we're still in trouble, as the
+subtractions in the last two lines can, as far as the symbolic
+simulator is concerned, introduce any value. So, supposing we're part
+way through the algorithm and we know that 3 most significant bits of
+`x` are `False` and the 5 most significant bis of `y` are `False`, but
+`x` and `y` are both odd; then we'll compute either `x - y` or `y - x`
+and the symbolic simulator will think we have no knowledge of the bits
+of either difference. The subsequent `>> 1` will introduce a single
+`False` bit to our computation, but overall we will have decreased the
+certainty of the bits in our arguments as we recurse.
+
+The next trick is to explicitly remove the unnecessary bits from the
+arguments as we recurse. If we represent halving by trimming a bit
+from the argument, the symbolic simulator reckons that one or both of
+the arguments eventually have type `[0]` which serves as a base
+case. What we have done is change from reducing values to eventually
+reach a base case, to reducing the types until the base case is
+reached. After that it's all details, but there are several.
+
+
+Implementing Binary gcd in Cryptol
+----------------------------------
+
+We'll need a few supporting functions:
+
++ A test for even is needed for various sized values. It is
+  implemented in terms of `odd` below.
+
+```cryptol
+/**
+ * Indicates whether the argument is even or odd.
+ */
+
+even, odd : {n} (fin n) => [n] -> Bit
+even = ~odd
+odd x = last ([False] # x) // "[False] #" affords type [0]
```

Comment:
How about
```
odd x = x!0
```
or even
```
odd (_#[b]) = b
```
? That `[False] # x` is...it's just...odd.

Reply:
It's necessary to keep that pesky n >= 1 constraint out.